File: 543. Diameter of Binary Tree/solution.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
  def diameterOfBinaryTree(self, root):
    """
    :type root: TreeNode
    :rtype: int
    """
    self.result = 0
    def dfs(current):
      if current == None:
        return 0
      left = dfs(current.left)
      right = dfs(current.right)
      
      if left + right > self.result:
        self.result = left + right

      maximum = left
      if right > maximum:
        maximum = right
      return maximum + 1
    
    dfs(root)
    return self.result

File: 543. Diameter of Binary Tree/solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
  int result = 0;
  int dfs(TreeNode current){
    if (current == null){
      return 0;
    }
    int left = dfs(current.left);
    int right = dfs(current.right);
    
    if (left + right > result){
      result = left + right;
    }

    int maximum = left;
    if (right > maximum){
      maximum = right;
    }
    return maximum + 1;
  }
  public int diameterOfBinaryTree(TreeNode root) {
    dfs(root);
    return result;
  }
}